### PR TITLE
Bazel workspace configuration support for running a flutter doctor script

### PR DESCRIPTION
**Pull request by jwren**
@pq @devoncarew

### CODE REVIEW

File: src/io/flutter/actions/FlutterSdkAction.java [hunk 1]
```diff
@@ -30,19 +32,36 @@
   public void actionPerformed(AnActionEvent event) {
     final Project project = DumbAwareAction.getEventProject(event);
 
-    final FlutterSdk sdk = project != null ? FlutterSdk.getFlutterSdk(project) : null;
-    if (sdk == null) {
-      showMissingSdkDialog(project);
-      return;
-    }
+    if(enableActionInBazelContext() && project != null) {
```
Comment (pq) [inline src/io/flutter/actions/FlutterSdkAction.java]: Looks like we're missing a space after `if`?  Perhaps re-format?

File: src/io/flutter/actions/FlutterDoctorAction.java [hunk 2]
```diff
@@ -5,15 +5,56 @@
...
+      FlutterConsoles.displayProcessLater(handler, project, null, handler::startNotify);
+    }
+    catch (ExecutionException e) {
+      LOG.error(e);
+    }
+  }
+
+  @Override
+  public boolean enableActionInBazelContext() {
+    return true;
```
Comment (pq) [inline src/io/flutter/actions/FlutterDoctorAction.java]: Doesn't this mean that `FlutterDoctorAction` will always run in the Bazel context? Doesn't this mean suggest that `startCommand` will never get run?

File: src/io/flutter/actions/FlutterSdkAction.java [hunk 3]
```diff
@@ -30,19 +32,36 @@
...
+      FlutterInitializer.sendAnalyticsAction(this);
+      FileDocumentManager.getInstance().saveAllDocuments();
+      startCommand(project, sdk, PubRoot.forEventWithRefresh(event));
+    }
   }
 
   public abstract void startCommand(@NotNull Project project, @NotNull FlutterSdk sdk, @Nullable PubRoot root);
 
+  public void startCommandInBazelContext(@NotNull Project project, @NotNull Workspace workspace) {
+  }
```
Comment (pq) [inline src/io/flutter/actions/FlutterSdkAction.java]: Maybe make this abstract?  Or if it's only appropriate for `FlutterDoctorAction`s, consider putting the conditional logic in there?

File: src/io/flutter/actions/FlutterSdkAction.java (hunk 1 above)
Comment (jwren) [inline src/io/flutter/actions/FlutterSdkAction.java]: Re-formatted

File: src/io/flutter/actions/FlutterDoctorAction.java (hunk 2 above)
Comment (jwren) [inline src/io/flutter/actions/FlutterDoctorAction.java]: Currently yes, but that is a bug... Fixing up for the PR.

File: src/io/flutter/actions/FlutterSdkAction.java (hunk 3 above)
Comment (jwren) [inline src/io/flutter/actions/FlutterSdkAction.java]: Not all actions can/should be run in the bazel context.  The idea here is to have methods override this method and enableActionInBazelContext if they should have the opportunity to run in this mode.

Added this commment to startCommandInBazelContext:


  /**
   * Implemented by actions which are used in the Bazel context ({@link #enableActionInBazelContext()} returns true), by default this method
   * is a no-op.
   */

**Comment by jwren** [comment]
PTAL

File: src/io/flutter/actions/FlutterSdkAction.java [hunk 4]
```diff
@@ -30,19 +32,43 @@
...
+    }
   }
 
   public abstract void startCommand(@NotNull Project project, @NotNull FlutterSdk sdk, @Nullable PubRoot root);
 
+  /**
+   * Implemented by actions which are used in the Bazel context ({@link #enableActionInBazelContext()} returns true), by default this method
+   * is a no-op.
+   */
+  public void startCommandInBazelContext(@NotNull Project project, @NotNull Workspace workspace) {
```
Comment (pq) [inline src/io/flutter/actions/FlutterSdkAction.java]: How about throwing an exception?  It seems like calling the no-op version would be a logic error?

File: src/io/flutter/actions/FlutterSdkAction.java [hunk 5]
```diff
@@ -30,19 +32,43 @@
   public void actionPerformed(AnActionEvent event) {
     final Project project = DumbAwareAction.getEventProject(event);
 
-    final FlutterSdk sdk = project != null ? FlutterSdk.getFlutterSdk(project) : null;
-    if (sdk == null) {
-      showMissingSdkDialog(project);
-      return;
+    // See if the Bazel workspace exists for this project.
+    final Workspace workspace = project != null ? WorkspaceCache.getInstance(project).getNow() : null;
```
Comment (pq) [inline src/io/flutter/actions/FlutterSdkAction.java]: Nit: don't we only need to access the workspace if `enableActionInBazelContext` returns true?

File: src/io/flutter/actions/FlutterDoctorAction.java (hunk 2 above)
Comment (pq) [inline src/io/flutter/actions/FlutterDoctorAction.java]: 👍

File: src/io/flutter/actions/FlutterSdkAction.java (hunk 4 above)
Comment (jwren) [inline src/io/flutter/actions/FlutterSdkAction.java]: I like that idea, done.

File: src/io/flutter/actions/FlutterSdkAction.java (hunk 5 above)
Comment (jwren) [inline src/io/flutter/actions/FlutterSdkAction.java]: Correct, a small optimization, but an optimization non-the-less.  Done.

**Pull request merged**